Support TOTP

AStove opened the pull request:
-Add checkbox to use TOTP
-This checkbox unhides TOTP input field
-This checkbox disable autoconnect
-Contents of TOTP field is sent to weechat
-Added usage instructions

Thoughts on sanitizing the TOTP field? It's always a 6 digit int (with leading zero's), should we enforce that? With full validation making the inputfield red etc?

Techman commented on js/connection.js:
```diff
@@ -328,7 +330,8 @@ weechat.factory('connection',
         $log.info('Attempting to reconnect...');
         var d = connectionData;
-        connect(d[0], d[1], d[2], d[3], d[4], function() {
+        // won't work if totp is mandetory
```
Slight spelling error here, correct spelling is 'mandatory'.

AStove replied:
ty

lorenzhs commented on index.html:
```diff
@@ -123,18 +123,26 @@ <h4 class="panel-title">
-                      <input type="checkbox" id="ssl" ng-model="settings.ssl">
-                      Encryption. <strong>Strongly recommended!</strong> Need help? Check below.
+                    <label class="control-label" for="useTotp">
+                      <input type="checkbox" id="useTotp" ng-model="settings.useTotp" ng-change="changeUseTOTP()">
```
you know you can add a callback for when a setting changes, too? https://github.com/glowing-bear/glowing-bear/blob/3d042b125bd40f51f4d9cc53fa9abb801f11cd71/js/settings.js#L52-L65

lorenzhs replied:
Using that is a bit cheaper on the CPU and more robust because it also triggers if any other code changes the setting, not just when the checkbox is changed

AStove replied:
I did not know, thanks. Changed.

lorenzhs commented on index.html:
```diff
@@ -123,18 +123,26 @@ <h4 class="panel-title">
-                      Encryption. <strong>Strongly recommended!</strong> Need help? Check below.
+                    <label class="control-label" for="useTotp">
+                      <input type="checkbox" id="useTotp" ng-model="settings.useTotp" ng-change="changeUseTOTP()">
+                      Use Time-based One-Time Password
```
This probably needs an :information_source: link next to it to the weechat docs or some kind of explanation

lorenzhs commented:
Thanks, you've implemented a lot more of the things that were missing than I did :)

I kinda prefer the design of #1040 though, where the token input is next to the password input like host and port. This makes the dynamic switching a bit more complicated (need to add/remove classes), but what do you think?

Also, I don't have a test instance at hand right now so I can't check, but how do you handle reconnects? I guess they should simply be disabled?

lorenzhs commented on index.html:
```diff
@@ -123,18 +123,26 @@ <h4 class="panel-title">
                     </label>
                   </div>
+                  <label class="control-label" for="totp" ng-Show="settings.useTotp">One time password</label>
+                  <input type="text" class="form-control favorite-font" id="totp" ng-model="totp" ng-Show="settings.useTotp" ng-model="totp" autocomplete="off">
```
you've specified `ng-model` twice here

AStove commented:
-Changed the layout to match #1040, also separated the port and hostname label this seemed fitting.
-Added information link next to totp checkbox
-Don't attempt reconnect when useTotp setting is used
-Validate totp

Loosely related: I noticed weechat accepts totp argument even when totp is not enabled. I made a [feature request](https://github.com/weechat/weechat/issues/1435) to this for weechat.

lorenzhs commented on index.html:
```diff
@@ -123,20 +133,26 @@ <h4 class="panel-title">
-                      Encryption. <strong>Strongly recommended!</strong> Need help? Check below.
+                    <label class="control-label" for="useTotp">
+                      <input type="checkbox" id="useTotp" ng-model="settings.useTotp">
+                      Use Time-based One-Time Password <a href="https://blog.weechat.org/post/2019/01/14/Support-of-TOTP">ℹ️</a>
```
Link should be `target="_blank"` I guess, but this applies to the other links as well, so we should just change it everywhere at some point

lorenzhs replied:
Also maybe more prominent. Bootstrap comes with some icons, maybe use those:
```html
<i class="glyphicon glyphicon-info-sign"></i>
```

lorenzhs commented on index.html:
```diff
@@ -123,20 +133,26 @@ <h4 class="panel-title">
+                      <input type="checkbox" id="ssl" ng-model="settings.ssl">
+                      Encryption. <strong>Strongly recommended!</strong> Need help? Check below.
+                    </label>
+                  </div>
```
Why reorder the inputs like that? It's weird to have save password -> encryption -> autoconnect when autoconnect depends on save password

AStove replied:
Agree but I didn't specifically reorder. I initially added it to the bottom and then moved the input to the top.. Better?

lorenzhs commented on js/glowingbear.js:
```diff
@@ -687,6 +688,16 @@ weechat.controller('WeechatCtrl', ['$rootScope', '$scope', '$store', '$timeout',
+    });
+
+    $scope.parseTotp = function() {
+        $scope.totpInvalid = !/^\d{6}$/.test($scope.totp);
```
From the linked weechat blog post:

> the number of digits (4 to 10 digits are supported, 6 is the default and recommended value).

lorenzhs commented:
Thanks, I really like this :)

lorenzhs commented:
Thanks a lot!
